cmd/manualchange.go

```diff
@@ -0,0 +1,388 @@
+	// nolint: bodyclose // nhooyr.io/websocket reads the body internally
+	c, _, err := websocket.Dial(ctx, viper.GetString("gateway-url"), options)
+	if err != nil {
+		log.WithContext(ctx).WithFields(lf).WithError(err).Error("Failed to connect to overmind API")
+		return 1
+	}
+	defer c.Close(websocket.StatusGoingAway, "")
+
+	// the default, 32kB is too small for cert bundles and rds-db-cluster-parameter-groups
+	c.SetReadLimit(2 * 1024 * 1024)
+
+	queriesSentChan := make(chan struct{})
+	go func() {
+		for _, q := range queries {
+			req := sdp.GatewayRequest{
+				MinStatusInterval: minStatusInterval,
+				RequestType: &sdp.GatewayRequest_Query{
+					Query: q,
+				},
+			}
+			err = wspb.Write(ctx, c, &req)
+
+			if err == nil {
+				log.WithContext(ctx).WithFields(log.Fields{
+					"scope":  q.Scope,
+					"type":   q.Type,
+					"query":  q.Query,
+					"method": q.Method.String(),
+					"uuid":   q.ParseUuid().String(),
+				}).Trace("Started query")
+			}
+			if err != nil {
+				log.WithContext(ctx).WithFields(lf).WithError(err).WithField("req", &req).Error("Failed to send request")
+				continue
+			}
+		}
+		queriesSentChan <- struct{}{}
+	}()
+
+	responses := make(chan *sdp.GatewayResponse)
+
+	// Start a goroutine that reads responses
+	go func() {
+		for {
+			res := new(sdp.GatewayResponse)
+
+			err = wspb.Read(ctx, c, res)
+
+			if err != nil {
+				var e websocket.CloseError
+				if errors.As(err, &e) {
+					log.WithContext(ctx).WithFields(lf).WithFields(log.Fields{
+						"code":   e.Code.String(),
+						"reason": e.Reason,
+					}).Info("Websocket closing")
+					return
+				}
+				log.WithContext(ctx).WithFields(lf).WithError(err).Error("Failed to read response")
+				return
+			}
+
+			responses <- res
+		}
+	}()
+
+	activeQueries := make(map[uuid.UUID]bool)
+	queriesSent := false
+
+	// Read the responses
+responses:
+	for {
+		select {
+		case <-queriesSentChan:
+			queriesSent = true
+
+		case <-signals:
+			log.WithContext(ctx).WithFields(lf).Info("Received interrupt, exiting")
+			return 1
+
+		case <-ctx.Done():
+			log.WithContext(ctx).WithFields(lf).Info("Context cancelled, exiting")
+			return 1
+
+		case resp := <-responses:
+			switch resp.ResponseType.(type) {
+
+			case *sdp.GatewayResponse_Status:
+				status := resp.GetStatus()
+				statusFields := log.Fields{
+					"summary":                  status.Summary,
+					"responders":               status.Summary.Responders,
+					"queriesSent":              queriesSent,
+					"post_processing_complete": status.PostProcessingComplete,
+				}
+
+				if status.Done() {
+					// fall through from all "final" query states, check if there's still queries in progress;
+					// only break from the loop if all queries have already been sent
+					// TODO: see above, still needs DefaultStartTimeout implemented to account for slow sources
+					allDone := allDone(ctx, activeQueries, lf)
```

Comment:
Realistically this check shouldn't be required as the gateway shouldn't tell us we're done unless *everything* is done right? Can we change maybe the "all responders done, with unfinished queries" log to an error so it's easier to spot? Ideally if this every happens we'd know about it

Reply:
`allDone` has a `Debugf("query still active")` for exactly that.